code/obj/machinery/manufacturer.dm

```diff
@@ -1026,44 +1026,47 @@ TYPEINFO(/obj/machinery/manufacturer)
 		storage[index_2] = temp_hold
 		src.should_update_static = TRUE
 
-	proc/eject_material(var/mat_ref, mob/user = null)
-		if (src.mode != MODE_READY)
-			src.grump_message(user, "You cannot eject materials while the unit is working.")
-			return
-		var/ejectamt = 0
-		var/turf/ejectturf = get_output_location(user)
-		var/obj/item/material_piece/target
-		for(var/obj/item/material_piece/P in src.get_contents())
-			if ("\ref[P]" == mat_ref)
-				target = P
-				break
-		if (isnull(target))
-			src.grump_message(user, "ERROR: Material not found in storage.", sound = TRUE)
+	proc/eject_material(var/mat_ref, mob/user)
+		var/obj/item/material_piece/P = src.get_material_by_ref(mat_ref)
+		if (!src.can_eject_material(P, user))
 			return
-		if (target.amount < 1)
-			src.grump_message(user, "ERROR: Not enough material to eject bars.", sound = TRUE)
+		var/eject_amount = tgui_input_number(user,
+											 "How many material pieces do you want to eject?",
+											 title = "Eject Materials",
+											 default = 0,
+											 max_value = ceil(P.amount),
+											 min_value = 0,
+											 round_input = TRUE,
+											)
+		if (!src.can_eject_material(P, user))
 			return
-		// Since tgui_input_number() is awaited here, make sure they are adjacent before and after
-		if (!src.check_physical_proximity(user))
+		// In case tgui_input_number() misses something or it's 0 (Likely trying to cancel action)
+		if (!isnum(eject_amount) || !isnum_safe(eject_amount) || eject_amount < 1)
```

Comment:
Do we need to check for fractional amounts, e.g. `1.5`?

Reply:
No, tgui_input_number rounds this